soroban-sdk/src/events.rs

```diff
@@ -0,0 +1,118 @@
+use core::fmt::Debug;
+
+#[cfg(doc)]
+use crate::{contracttype, Bytes, BytesN, Map};
+use crate::{env::internal, Env, IntoVal, Object, RawVal, TryFromVal, Vec};
+
+// TODO: consolidate with host::events::TOPIC_BYTES_LENGTH_LIMIT
+const TOPIC_BYTES_LENGTH_LIMIT: u32 = 32;
+
+/// Events publishes events for the currently executing contract.
+///
+/// ```
+/// use soroban_sdk::Env;
+///
+/// # use soroban_sdk::{contractimpl, vec, map, RawVal, BytesN};
+/// #
+/// # pub struct Contract;
+/// #
+/// # #[contractimpl]
+/// # impl Contract {
+/// #     pub fn f(env: Env) {
+/// let event = env.events();
+/// let data = map![&env, (1u32, 2u32)];
+/// let topics0 = ();
+/// let topics1 = (0u32,);
+/// let topics2 = (0u32, 1u32);
+/// let topics3 = (0u32, 1u32, 2u32);
+/// let topics4 = (0u32, 1u32, 2u32, 3u32);
+/// let topics_vec = vec![&env, 4u32, 5u32, 6u32, 7u32];
+/// event.publish(topics0, data.clone());
+/// event.publish(topics1, data.clone());
+/// event.publish(topics2, data.clone());
+/// event.publish(topics3, data.clone());
+/// event.publish(topics4, data.clone());
+/// event.publish(topics_vec, data.clone());
+/// #     }
+/// # }
+///
+/// # #[cfg(feature = "testutils")]
+/// # fn main() {
+/// #     let env = Env::default();
+/// #     let contract_id = BytesN::from_array(&env, [0; 32]);
+/// #     env.register_contract(&contract_id, Contract);
+/// #     f::invoke(&env, &contract_id);
+/// # }
+/// # #[cfg(not(feature = "testutils"))]
+/// # fn main() { }
+/// ```
+
+#[derive(Clone)]
+pub struct Events(Env);
+
+impl Debug for Events {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "Events")
+    }
+}
+
+pub trait Topics: IntoVal<Env, Object> {}
```

Comment:
I think this should be Vec<T>. That's ultimately what we're needing all the values to be. It'd mean adding some functions to tuples I think here in the SDK. Wdyt?
```suggestion
pub trait Topics: IntoVal<Env, Vec<T>> {}
```

Reply:
I'm trying it out but I think there is a difference. 
Ultimately what goes into the host a `Vec`, i.e., heterogeneous vector, which tuple `(T1, T2, T3, T4)` gets converted to. 
But `Vec<T>` implies this must be a homogenous typed vector. This goes back to the issue at the beginning and why this PR started. 
I'll experiment a little bit to see if there's an elegant way, if not might have to fall back to the `IntoTopics` solution.

Reply:
Oh sorry, I should have said `Vec<RawVal>`. I think all these types are convertible into that?

Reply:
The main reason to do this is to support https://github.com/stellar/rs-soroban-sdk/pull/469/files#r953184081.

Reply:
Ah yes, that should work.

Reply:
I changed the trait to `IntoVal<Env, Vec<RawVal>>` as you suggested. It works fairly nicely for the tuples. 
However, I found it much less elegant to work with `Vec<T>` since it requires 
```
impl<T> IntoVal<Env, Vec<RawVal>> for Vec<T>
where
    T: IntoVal<Env, RawVal> + TryFromVal<Env, RawVal>,
```
which ironically I found it hard to do. Ideally it should just do something like `self.iter().map(f).collect()` into `Vec<RawVal>` but that didn't work (or maybe I just didn't spend enough time).
@leighmcculloch If you are ok with it, I think we should merge this as is (since we are done with Tuples) and we can add a followup issue to support vector.

Reply:
Agreed adding vector was just a nicety. 

Fwiw to make it work I think you need a `impl<T> From<Vec<T>> for Vec<RawVal> where T: IntoVal<Env, RawVal>` and inside that conversion you just do an unsafe unchecked_new since we can trust the types are RawVal.